File: tools/hydra-ts/test/client.test.ts
import { describe, it, before, after } from 'node:test';
import assert from 'node:assert/strict';
import http from 'node:http';
import type { AddressInfo } from 'node:net';
import { HydraClient, type HydraClientConfig } from '../src/client.js';

interface TestNode {
  leaderStatus: number;
  endpointStatus: number;
  endpointCalls: number;
  auth?: string;
}

const nodes = new Map<string, TestNode>();
const servers: http.Server[] = [];

function startNode(initial: Partial<TestNode> = {}): Promise<string> {
  const state: TestNode = {
    leaderStatus: 503,
    endpointStatus: 200,
    endpointCalls: 0,
    ...initial,
  };
  return new Promise((resolve) => {
    const server = http.createServer((req, res) => {
      const url = new URL(req.url ?? '/', 'http://localhost');
      if (url.pathname === '/healthz/leader') {
        res.writeHead(state.leaderStatus, { 'Content-Type': 'application/json' });
        res.end(state.leaderStatus === 200 ? '{"leader":true}' : '');
        return;
      }
      if (url.pathname === '/api/v1/tenants/t-acme/auth/cache/invalidate' && req.method === 'POST') {
        state.endpointCalls += 1;
        state.auth = req.headers.authorization;
        res.writeHead(state.endpointStatus, { 'Content-Type': 'application/json' });
        res.end(state.endpointStatus >= 200 && state.endpointStatus < 300 ? '{"invalidated":1}' : '');
        return;
      }
      res.writeHead(404);
      res.end();
    });
    server.listen(0, '127.0.0.1', () => {
      const { port } = server.address() as AddressInfo;
      nodes.set(`http://127.0.0.1:${port}`, state);
      servers.push(server);
      resolve(`http://127.0.0.1:${port}`);
    });
  });
}

function makeClient(urls: string[], extra: Partial<HydraClientConfig> = {}) {
  return new HydraClient({
    token: 'sk-tenant-token',
    nodes: urls,
    probeTimeoutMs: 1000,
    requestTimeoutMs: 1000,
    disableBackgroundRecheck: true,
    ...extra,
  });
}

describe('HydraClient', () => {
  before(async () => {
    // no-op; servers are created per test and tracked below
  });
  after(() => {
    for (const server of servers) server.close();
  });

  it('uses the leader even when it is not first', async () => {
    const standby = await startNode({ leaderStatus: 503 });
    const leader = await startNode({ leaderStatus: 200 });
    const client = makeClient([standby, leader]);
    client.close();
    await client.invalidateTenantAuthCache('t-acme');
    assert.equal(nodes.get(leader)!.endpointCalls, 1);
    assert.equal(nodes.get(standby)!.endpointCalls, 0);
  });

  it('fails over and removes dead nodes', async () => {
    const dead = await startNode({ leaderStatus: 200, endpointStatus: 500 });
    const healthy = await startNode({ leaderStatus: 503 });
    const client = makeClient([dead, healthy]);
    client.close();
    await client.invalidateTenantAuthCache('t-acme');
    assert.equal(nodes.get(dead)!.endpointCalls, 1);
    assert.equal(nodes.get(healthy)!.endpointCalls, 1);
    assert.deepEqual(client.nodes, [healthy]);
    assert.deepEqual(client.removedNodes, [dead]);
  });

  it('probeRemovedNodes restores reachable nodes', async () => {
    const dead = await startNode({ leaderStatus: 200, endpointStatus: 500 });
    const healthy = await startNode({ leaderStatus: 503 });
    const client = makeClient([dead, healthy]);
    client.close();
    await client.invalidateTenantAuthCache('t-acme');
    assert.equal(client.removedNodes.length, 1);
    nodes.get(dead)!.endpointStatus = 200;
    await client.probeRemovedNodes();
    assert.deepEqual(new Set(client.nodes), new Set([dead, healthy]));
    assert.deepEqual(client.removedNodes, []);
  });

  it('supports single node without leader probe', async () => {
    const node = await startNode({ leaderStatus: 404 });
    const client = makeClient([node]);
    client.close();
    await client.invalidateTenantAuthCache('t-acme');
    assert.equal(nodes.get(node)!.endpointCalls, 1);
  });

  it('does not remove nodes on 401', async () => {
    const node = await startNode({ leaderStatus: 200, endpointStatus: 401 });
    const client = makeClient([node]);
    client.close();
    await assert.rejects(() => client.invalidateTenantAuthCache('t-acme'), /401/);
    assert.deepEqual(client.removedNodes, []);
    assert.deepEqual(client.nodes, [node]);
  });

  it('sends Authorization and api_keys body', async () => {
    let seenAuth: string | undefined;
    let seenBody = '';
    const server = http.createServer((req, res) => {
      if (req.url === '/healthz/leader') {
        res.writeHead(200);
        res.end();
        return;
      }
      let body = '';
      req.on('data', (chunk) => { body += chunk; });
      req.on('end', () => {
        if (req.url === '/api/v1/tenants/t-acme/auth/cache/invalidate' && req.method === 'POST') {
          seenAuth = req.headers.authorization;
          seenBody = body;
        }
        res.writeHead(200);
        res.end('{"invalidated":1}');
      });
    });
    await new Promise<void>((resolve) => server.listen(0, '127.0.0.1', resolve));
    servers.push(server);
    const { port } = server.address() as AddressInfo;
    const client = makeClient([`http://127.0.0.1:${port}`]);
    client.close();
    await client.invalidateTenantAuthCacheKeys('t-acme', ['key1', 'key2']);
    assert.equal(seenAuth, 'Bearer sk-tenant-token');
    assert.match(seenBody, /key1/);
    assert.match(seenBody, /key2/);
  });

  it('background recheck restores removed nodes', async () => {
    const dead = await startNode({ leaderStatus: 200, endpointStatus: 500 });
    const healthy = await startNode({ leaderStatus: 503 });
    const client = new HydraClient({
      token: 'sk-tenant-token',
      nodes: [dead, healthy],
      probeTimeoutMs: 1000,
      requestTimeoutMs: 1000,
      recheckIntervalMs: 10,
    });
    try {
      await client.invalidateTenantAuthCache('t-acme');
      nodes.get(dead)!.endpointStatus = 200;
      const deadline = Date.now() + 2000;
      while (Date.now() < deadline) {
        if (client.nodes.length === 2 && client.removedNodes.length === 0) return;
        await new Promise((resolve) => setTimeout(resolve, 10));
      }
      assert.fail(`node not restored: nodes=${client.nodes} removed=${client.removedNodes}`);
    } finally {
      client.close();
    }
  });
});

File: tools/hydra-go/client_test.go
package hydra

import (
	"context"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync"
	"sync/atomic"
	"testing"
	"time"
)

type testNode struct {
	mu       sync.Mutex
	leader   int // HTTP status for /healthz/leader
	endpoint int // HTTP status for the tenant invalidation endpoint
	calls    atomic.Int32
}

func newTestNode(leaderStatus, endpointStatus int) *testNode {
	return &testNode{leader: leaderStatus, endpoint: endpointStatus}
}

func (n *testNode) setEndpoint(status int) {
	n.mu.Lock()
	defer n.mu.Unlock()
	n.endpoint = status
}

func (n *testNode) handler(t *testing.T) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch {
		case r.URL.Path == "/healthz/leader":
			n.mu.Lock()
			status := n.leader
			n.mu.Unlock()
			if status == http.StatusOK {
				w.Header().Set("Content-Type", "application/json")
				w.WriteHeader(status)
				_, _ = w.Write([]byte(`{"leader":true}`))
				return
			}
			w.WriteHeader(status)
			return
		case r.URL.Path == "/api/v1/tenants/t-acme/auth/cache/invalidate" && r.Method == http.MethodPost:
			n.calls.Add(1)
			if auth := r.Header.Get("Authorization"); auth != "Bearer sk-tenant-token" {
				t.Errorf("unexpected Authorization header: %q", auth)
			}
			n.mu.Lock()
			status := n.endpoint
			n.mu.Unlock()
			if status >= 200 && status < 300 {
				w.Header().Set("Content-Type", "application/json")
				w.WriteHeader(status)
				_, _ = w.Write([]byte(`{"invalidated":1,"tenant_id":"t-acme"}`))
				return
			}
			w.WriteHeader(status)
			return
		default:
			http.NotFound(w, r)
		}
	})
}

func newClient(t *testing.T, nodes []string) *Client {
	t.Helper()
	c, err := New(Config{
		Token:                    "sk-tenant-token",
		Nodes:                    nodes,
		DisableBackgroundRecheck: true,
		ProbeTimeout:             time.Second,
		RequestTimeout:           time.Second,
	})
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	t.Cleanup(func() { _ = c.Close() })
	return c
}

func TestInvalidateUsesLeaderNotFirstNode(t *testing.T) {
	standby := newTestNode(http.StatusServiceUnavailable, http.StatusOK)
	leader := newTestNode(http.StatusOK, http.StatusOK)

	standbySrv := httptest.NewServer(standby.handler(t))
	defer standbySrv.Close()
	leaderSrv := httptest.NewServer(leader.handler(t))
	defer leaderSrv.Close()

	c := newClient(t, []string{standbySrv.URL, leaderSrv.URL})
	if err := c.InvalidateTenantAuthCache(context.Background(), "t-acme"); err != nil {
		t.Fatalf("InvalidateTenantAuthCache: %v", err)
	}
	if leader.calls.Load() != 1 {
		t.Fatalf("leader endpoint calls = %d, want 1", leader.calls.Load())
	}
	if standby.calls.Load() != 0 {
		t.Fatalf("standby endpoint calls = %d, want 0", standby.calls.Load())
	}
}

func TestInvalidateFailsOverAndRemovesDeadNode(t *testing.T) {
	dead := newTestNode(http.StatusOK, http.StatusInternalServerError)
	healthy := newTestNode(http.StatusServiceUnavailable, http.StatusOK)

	deadSrv := httptest.NewServer(dead.handler(t))
	defer deadSrv.Close()
	healthySrv := httptest.NewServer(healthy.handler(t))
	defer healthySrv.Close()

	c := newClient(t, []string{deadSrv.URL, healthySrv.URL})
	if err := c.InvalidateTenantAuthCache(context.Background(), "t-acme"); err != nil {
		t.Fatalf("InvalidateTenantAuthCache: %v", err)
	}
	if dead.calls.Load() != 1 {
		t.Fatalf("dead node endpoint calls = %d, want 1", dead.calls.Load())
	}
	if healthy.calls.Load() != 1 {
		t.Fatalf("healthy node endpoint calls = %d, want 1", healthy.calls.Load())
	}
	active := c.Nodes()
	if len(active) != 1 || active[0] != healthySrv.URL {
		t.Fatalf("active nodes = %v, want [%s]", active, healthySrv.URL)
	}
	removed := c.RemovedNodes()
	if len(removed) != 1 || removed[0] != deadSrv.URL {
		t.Fatalf("removed nodes = %v, want [%s]", removed, deadSrv.URL)
	}
}

func TestProbeRemovedNodesRestoresReachableNode(t *testing.T) {
	node := newTestNode(http.StatusOK, http.StatusInternalServerError)
	healthy := newTestNode(http.StatusServiceUnavailable, http.StatusOK)

	nodeSrv := httptest.NewServer(node.handler(t))
	defer nodeSrv.Close()
	healthySrv := httptest.NewServer(healthy.handler(t))
	defer healthySrv.Close()

	c := newClient(t, []string{nodeSrv.URL, healthySrv.URL})
	if err := c.InvalidateTenantAuthCache(context.Background(), "t-acme"); err != nil {
		t.Fatalf("first invalidate: %v", err)
	}
	if len(c.Nodes()) != 1 || len(c.RemovedNodes()) != 1 {
		t.Fatalf("before restore active=%v removed=%v", c.Nodes(), c.RemovedNodes())
	}

	// The node is reachable again; the next probe should add it back.
	node.setEndpoint(http.StatusOK)
	c.ProbeRemovedNodes(context.Background())

	active := c.Nodes()
	if len(active) != 2 {
		t.Fatalf("active nodes after restore = %v, want 2", active)
	}
	if len(c.RemovedNodes()) != 0 {
		t.Fatalf("removed nodes after restore = %v, want empty", c.RemovedNodes())
	}
}

func TestSingleNodeWithoutLeaderProbe(t *testing.T) {
	node := newTestNode(http.StatusNotFound, http.StatusOK)
	srv := httptest.NewServer(node.handler(t))
	defer srv.Close()

	c := newClient(t, []string{srv.URL})
	if err := c.InvalidateTenantAuthCache(context.Background(), "t-acme"); err != nil {
		t.Fatalf("InvalidateTenantAuthCache: %v", err)
	}
	if node.calls.Load() != 1 {
		t.Fatalf("endpoint calls = %d, want 1", node.calls.Load())
	}
}

func TestAuthErrorDoesNotRemoveNode(t *testing.T) {
	node := newTestNode(http.StatusOK, http.StatusUnauthorized)
	srv := httptest.NewServer(node.handler(t))
	defer srv.Close()

	c := newClient(t, []string{srv.URL})
	err := c.InvalidateTenantAuthCache(context.Background(), "t-acme")
	if err == nil {
		t.Fatal("expected auth error")
	}
	if !strings.Contains(err.Error(), "401") {
		t.Fatalf("error = %v, want 401", err)
	}
	if len(c.RemovedNodes()) != 0 {
		t.Fatalf("removed nodes = %v, want empty", c.RemovedNodes())
	}
	if len(c.Nodes()) != 1 {
		t.Fatalf("active nodes = %v, want 1", c.Nodes())
	}
}

func TestInvalidateTenantAuthCacheKeysSendsBody(t *testing.T) {
	var gotBody string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch {
		case r.URL.Path == "/healthz/leader":
			w.WriteHeader(http.StatusOK)
		case r.URL.Path == "/api/v1/tenants/t-acme/auth/cache/invalidate" && r.Method == http.MethodPost:
			buf := make([]byte, 4096)
			n, _ := r.Body.Read(buf)
			gotBody = string(buf[:n])
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"invalidated":1}`))
		default:
			http.NotFound(w, r)
		}
	}))
	defer srv.Close()

	c := newClient(t, []string{srv.URL})
	if err := c.InvalidateTenantAuthCacheKeys(context.Background(), "t-acme", []string{"key1", "key2"}); err != nil {
		t.Fatalf("InvalidateTenantAuthCacheKeys: %v", err)
	}
	if !strings.Contains(gotBody, `"key1"`) || !strings.Contains(gotBody, `"key2"`) {
		t.Fatalf("body = %q, want api_keys with key1/key2", gotBody)
	}
}

func TestBackgroundRecheckRestoresRemovedNode(t *testing.T) {
	node := newTestNode(http.StatusOK, http.StatusInternalServerError)
	healthy := newTestNode(http.StatusServiceUnavailable, http.StatusOK)

	nodeSrv := httptest.NewServer(node.handler(t))
	defer nodeSrv.Close()
	healthySrv := httptest.NewServer(healthy.handler(t))
	defer healthySrv.Close()

	c, err := New(Config{
		Token:           "sk-tenant-token",
		Nodes:           []string{nodeSrv.URL, healthySrv.URL},
		ProbeTimeout:    time.Second,
		RequestTimeout:  time.Second,
		RecheckInterval: 10 * time.Millisecond,
	})
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	defer c.Close()

	if err := c.InvalidateTenantAuthCache(context.Background(), "t-acme"); err != nil {
		t.Fatalf("invalidate: %v", err)
	}
	if len(c.Nodes()) != 1 || len(c.RemovedNodes()) != 1 {
		t.Fatalf("before recovery active=%v removed=%v", c.Nodes(), c.RemovedNodes())
	}

	node.setEndpoint(http.StatusOK)
	deadline := time.Now().Add(2 * time.Second)
	for time.Now().Before(deadline) {
		if len(c.Nodes()) == 2 && len(c.RemovedNodes()) == 0 {
			return
		}
		time.Sleep(10 * time.Millisecond)
	}
	t.Fatalf("node was not restored: active=%v removed=%v", c.Nodes(), c.RemovedNodes())
}
